### PR TITLE
Auth0audit

### PR DESCRIPTION
**Pull request by siteshjha11**
Required items, please complete
   
   Change(s):
   - See guidance below

   Reason for Change(s):
   - See guidance below

   Version Updated:
   - Required only for Detections/Analytic Rule templates
   - See guidance below

   Testing Completed:
   - See guidance below

   Checked that the validations are passing and have addressed any issues that are present:
   - See guidance below

# Guidance <- remove section before submitting
-----------------------------------------------------------------------------------------------------------
## Before submitting this PR please ensure that you have read the following sections and filled out the changes, reason for change and testing complete sections:

Thank you for your contribution to the Microsoft Sentinel Github repo.

> Details of the code changes in your submitted PR.  Providing descriptions for pull requests ensures there is context to changes being made and greatly enhances the code review process.  Providing associated Issues that this resolves also easily connects the reason.
   
   Change(s):
   - Updated syntax for XYZ.yaml

   Reason for Change(s):
   - New schema used for XYZ.yaml
   - Resolves ISSUE #1234

   Version updated:
   - Yes
   - Detections/Analytic Rule templates are required to have the version updated

> The code should have been tested in a Microsoft Sentinel environment that does not have any custom parsers, functions or tables, so that you validate no incorrect syntax and execution functions properly.  If your submission requires a custom parser or function, it must be submitted with the PR.

   Testing Completed:
   - Yes/No/Need Help

_Note: If updating a detection, you must update the version field._

> Before the submission has been made, please look at running the KQL and Yaml Validation Checks locally.
> https://github.com/Azure/Azure-Sentinel#run-kql-validation-locally

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes/No/Need Help
   
   _Note: Let us know if you have tried fixing the validation error and need help._

> References: 
> - [Guidance for Detection checks](https://github.com/Azure/Azure-Sentinel#pull-request-detection-template-structure-validation-check)
> - [General contribution guidance](https://github.com/Azure/Azure-Sentinel/wiki#what-can-you-contribute-and-how-can-you-create-contributions)
> - [PR validation troubleshooting](https://github.com/Azure/Azure-Sentinel#pull-request)

-----------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: ASIM/dev/ASimTester.csv [hunk 1]
```diff
@@ -0,0 +1,1377 @@
+ColumnName,ColumnType,Class,Schema,LogicalType,ListOfValues,Aliased
```
Comment (vakohl) [inline ASIM/dev/ASimTester.csv:1]: incorrect location, please remove this change

File: Parsers/ASimAuditEvent/Parsers/Okta_Auth0_ASimAuditEvent.yaml [hunk 2]
```diff
@@ -0,0 +1,125 @@
+Parser:
+  Title: Audit Event ASIM parser for Auth0
+  Version: '0.1.0'
+  LastUpdated: Jun 13 2023
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/Okta_Auth0_ASimAuditEvent.yaml:4]: format incorrect. Also old date updated

File: Sample Data/ASIM/Okta_Auth0_AuditEvent_vimIngestedLogs.csv [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿"TimeGenerated [UTC]",EventCount,"EventStartTime [UTC]","EventEndTime [UTC]",EventProduct,EventVendor,EventSchema,EventSchemaVersion,EventUid,DvcFQDN,DvcIpAddr,DvcDomainType,DvcId,DvcAction,EventMessage,EventOriginalResultDetails,EventProductVersion,ActorUsername,ActorScope,ActorUserIdType,EventOriginalUid,Object,ObjectType,TargetIpAddr,Src,HttpUserAgent,TargetAppName,TargetAppType,DvcIdType,Type,EventType,EventSeverity,EventResult,Dvc,DvcHostname,Dst,SrcIpAddr,IpAddr,DvcDomain,Application,Operation,ActorUsernameType,User
```
Comment (vakohl) [inline Sample Data/ASIM/Okta_Auth0_AuditEvent_vimIngestedLogs.csv:1]: please attach results from table

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 14 records (100.0%) for field [ActorUsernameType] of type [Enumerated]: [""Oktaid""] (Schema:AuditEvent)"
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:2]: ActorUsernameType should be 'UPN'

image

Comment (vakohl) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:2]: image

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/Okta_Auth0_ASimAuditEvent.yaml (hunk 2 above)
Comment (rsingh134) [reply Parsers/ASimAuditEvent/Parsers/Okta_Auth0_ASimAuditEvent.yaml]: File is outdated.

File: ASIM/dev/ASimTester.csv (hunk 1 above)
Comment (rsingh134) [reply ASIM/dev/ASimTester.csv:1]: Please confirm what exactly required to be changed here.

File: Parsers/ASimAuditEvent/Parsers/Okta_Auth0_vimAuditEvent.yaml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+References:
+  - Title: Using functions
+    Link: https://docs.microsoft.com/azure/azure-monitor/log-query/function
+  - Title: ASIM Authentication Schema
+    Link: https://aka.ms/ASimAuthenticationDoc
+  - Title: ASIM
+    Link: https:/aka.ms/AboutASIM
+  - Title: Auth0
+    Link: https://auth0.com/docs/deploy-monitor/logs/log-event-type-codes
+Description: |
```
[the remark this answers is not in the document]
Comment (rsingh134) [reply Parsers/ASimAuditEvent/Parsers/Okta_Auth0_vimAuditEvent.yaml]: File is outdated!

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+              DvcAction=scope_s,         
+              EventMessage=description_s,         
+              EventOriginalResultDetails=type_s,
+              EventProductVersion='4.3.1',
+              ActorUsername=user_name_s,
+              ActorUsernameType='UPN',
+              ActorScope=tolower(user_name_s),
+              ActorUserIdType='Oktaid',
+              EventOriginalUid=_id_s,
+              Object=client_name_s,
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:148]: this should have Object detail. e.g.
For type_s 'sapi' and description_s 'Update Brute-force settings' Object would be 'Brute Force Setting' ObjectType as 'Configuration Atom'

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+              HttpUserAgent=user_agent_s,
+              TargetAppName="Auth0",
+              TargetAppType="Saas application",
+              DvcIdType='Other',      
+              Type
+              |lookup Auth0AuditEventLookup on EventOriginalResultDetails
+              | where      ( eventresult == '*' or EventResult =~eventresult)
+                   and((array_length(actorusername_has_any) == 0) or (ActorUsername  has_any (actorusername_has_any)))
+                   and ((array_length(eventtype_in) == 0) or (EventType in~ (eventtype_in)))
+                   and (array_length(newvalue_has_any) == 0) //new value not available at source
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:161]: we should check this newvalue filter early with SrcIp filter at line 124

File: Parsers/ASimAuthentication/Parsers/ASimAuthentication.yaml [hunk 8]
```diff
@@ -47,7 +47,8 @@ ParserQuery: |
     ASimAuthenticationGoogleWorkspace (ASimAuthenticationDisabled or ('ExcludeASimAuthenticationGoogleWorkspace' in (DisabledParsers) )),
```
Comment (vakohl) [inline Parsers/ASimAuthentication/Parsers/ASimAuthentication.yaml:47]: Please remove Authentication files from this PR

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            eventtype_in:dynamic=dynamic([]),
+            operation_has_any:dynamic=dynamic([]),
+            object_has_any:dynamic=dynamic([]),
+            newvalue_has_any:dynamic=dynamic([]),
+            disabled:bool = false
+            )
+     {
+     let Auth0AuditEventLookup=datatable(EventOriginalResultDetails:string,EventType:string,EventSeverity:string,EventResult:string)
+     [
+     "cls","other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:63]: EventMessage=description_s,   --> Populates if the table throws a description of the event. If we add it enumerated , we will have to keep few empty as they are not there . Please advise

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+EquivalentBuiltInParser: _ASim_AuditEvent_Auth0
+ParserParams:
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
+     let EventResultDetailslookup=datatable
+     ( 
+     EventOriginalResultDetails:string,
+     EventResult:string,
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml:29]: We only have these 2 error codes, is that sufficient to incorporate for all audit events?

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+     "cs","Other","Informational","Failure",
+     "depnote","Other","Informational","Failure",
+     "fc","Other","Informational","Failure",
+     "fce","Other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:67]: All Auth0 logs are around user management as the app itself is about authenticating users/machine etc  and very few with policies ,. If we remove these codes, we will only be left with 3-4 events as audit event. Please advise.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+     "fcp","Other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:68]: All Auth0 logs are around user management as the app itself is about authenticating users/machine etc and very few with policies ,. If we remove these codes, we will only be left with 3-4 events as audit event. Please advise.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+     "fcph","Other","Informational","Failure",
+     "fcpn","Other","Informational","Failure",
+     "fcpr","Other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:71]: All Auth0 logs are around user management as the app itself is about authenticating users/machine etc and very few with policies ,. If we remove these codes, we will only be left with 3-4 events as audit event. Please advise.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+     "fcpro","Other","Informational","Failure",
+     "fcu","Other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:73]: All Auth0 logs are around user management as the app itself is about authenticating users/machine etc and very few with policies ,. If we remove these codes, we will only be left with 3-4 events as audit event. Please advise.

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 14 records (100.0%) for field [EventVendor] of type [Enumerated]: [""Okta""] (Schema:AuditEvent)"
+"(1) Warning: Empty value in 14 records (100.0%)  in mandatory field [Dvc] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in optional field [DvcFQDN] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcAction] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcDomain] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcHostname] (Schema:AuditEvent)"
+"(2) Info: Empty value in 4 records (28.57%)  in optional field [EventMessage] (Schema:AuditEvent)"
+"(2) Info: Empty value in 6 records (42.86%)  in optional field [ActorScope] (Schema:AuditEvent)"
+"(2) Info: Empty value in 6 records (42.86%)  in recommended field [ActorUsername] (Schema:AuditEvent)"
+"(2) Info: Empty value in 6 records (42.86%)  in recommended field [Object] (Schema:AuditEvent)"
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:13]: EventMessage is mapped with description, the table does not have description of the events yet .. As mentioned, there are descriptions for few events at the source and some do not have . We only have secft and 2 more varities of events tested which do not have description.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+     "fd","Other","Informational","Failure",
+     "fdeac","Other","Informational","Failure",
+     "fdeaz","Other","Informational","Failure",
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:76]: All Auth0 logs are around user management as the app itself is about authenticating users/machine etc and very few with policies ,. If we remove these codes, we will only be left with 3-4 events as audit event. Please advise.

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv [hunk 17]
```diff
@@ -0,0 +1,15 @@
+﻿Result
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 14 records (100.0%) for field [ActorUsernameType] of type [Enumerated]: [""Oktaid""] (Schema:AuditEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 14 records (100.0%) for field [EventProduct] of type [Enumerated]: [""Auth0""] (Schema:AuditEvent)"
+"(0) Error: 1 invalid value(s) (up to 10 listed) in 14 records (100.0%) for field [EventVendor] of type [Enumerated]: [""Okta""] (Schema:AuditEvent)"
+"(1) Warning: Empty value in 14 records (100.0%)  in mandatory field [Dvc] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in optional field [DvcFQDN] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcAction] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcDomain] (Schema:AuditEvent)"
+"(2) Info: Empty value in 14 records (100.0%)  in recommended field [DvcHostname] (Schema:AuditEvent)"
+"(2) Info: Empty value in 4 records (28.57%)  in optional field [EventMessage] (Schema:AuditEvent)"
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:10]: Already mapped with decription_s , did you mean to enumerate it in data table?

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv (hunk 15 above)
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:13]: Object shall populate when an actual app to AUth0 authentication happens which will have a SAML provider like ping federate etc. The field has been mapped based on product schema .

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+         EventProduct="Auth0",
+         EventSchema="Authentication",
+         EventSchemaVersion='0.1.3',
+         EventUid=_ItemId,
+         EventStartTime=date_t,
+         EventEndTime=TimeGenerated,
+         TargetUsername=details_request_auth_user_name_s,
+         TargetUserId=details_request_auth_user_user_id_s,         
+         DvcFQDN=hostname_s,         
+         DvcIpAddr=ip_s,
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml:98]: ip_s is the Source DVC Ip addr which is Auth0 , while details request_s is the client that is trying to make use of Auth0. Please advise.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml (hunk 6 above)
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml:148]: Can incorporate, please confirm on the event description approach as mentioned in the erstwhile comments, do we want to make that enumerated or fetch it from the Auth0AM _CL table?

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml]: Please add it to the table. Later check if descrition_s is present only then overwrite EventMessage with descrition_s else keep it unchanged with value from datatable.

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv (hunk 17 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:10]: yes, explained in the another comment

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml]: details_request_ip_s as SrcIpAddr in that case.
DvcIpAddr = ip_s

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventAuth0.yaml.yaml]: two not enough. We sure it only provide two event code only?

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml (hunk 6 above)
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml]: we can evaluate EventMessage first using both datatable and description_s approach. And then set Object details based on EventMessage. This way Object won't be null when description_s is not available.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml]: Can you please list the ones with pure audit related activities, exclude UserManagement?

File: ASIM/dev/ASimTester.csv (hunk 1 above)
Comment (vakohl) [reply ASIM/dev/ASimTester.csv:1]: remove adding this file. The correct ASIMTester.csv is already being updated.

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv (hunk 15 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:13]: Object refers to: The name of the object on which the operation identified by [EventType](https://learn.microsoft.com/en-us/azure/sentinel/normalization-schema-audit#eventtype) is performed.

File: Parsers/ASimAuditEvent/Parsers/Okta_Auth0_vimAuditEvent.yaml [hunk 19]
```diff
@@ -0,0 +1,187 @@
+    Parser:
```
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/Okta_Auth0_vimAuditEvent.yaml]: Done . The file has been renamed.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventAuth0.yaml.yaml]: I feel the following might qualify, the MFA ones am unsure of , but listing them:

fcpro	Failed Connector Provisioning	Failed to provision a AD/LDAP connector
fd	Failed Delegation	Failed to generate delegation token
fdeac	Failed Device Activation	
fn	Failed Sending Notification	Failed to send email
fpar	Failed Pushed Authorization Request
fv	Failed Verification Email	Failed to send verification email
fvr	Failed Verification Email Request
gd_send_email	Email Sent	Email for MFA successfully sent.
gd_send_pn	Push notification sent	Push notification for MFA successfully sent.
gd_send_pn_failure	Push notification sent	Push notification for MFA failed.
gd_send_sms	SMS sent	SMS for MFA successfully sent.
gd_send_sms_failure	SMS sent failures	Attempt to send SMS for MFA failed.
gd_send_voice	Voice call made	Voice call for MFA successfully made.
gd_send_voice_failure	Voice call failure	Attempt to make Voice call for MFA failed.

gd_start_enroll	Enroll started	Multi-factor authentication enroll has started.
gd_start_enroll_failed	Enrollment failed	Push to start enrollement failed.
gd_tenant_update	Guardian tenant update	
gd_unenroll	Unenroll device account	Device used for second factor authentication has been unenrolled.
gd_update_device_account	Update device account	Device used for second factor authentication has been updated.
limit_delegation	Too Many Calls to /delegation	Rate limit exceeded to /delegation endpoint
mgmt_api_read	Management API read Operation	API GET operation returning secrets completed successfully
pla	Pre-login assessment	This log is generated before a login and helps in monitoring the behavior of bot detection without having to enable it.
srrt	Success Revocation	Successfully revoked a Refresh Token
sapi	Success API Operation

File: Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv (hunk 17 above)
[the remark this answers is not in the document]
Comment (siteshjha11) [reply Parsers/ASimAuditEvent/Tests/Okta_Auth0_ASimAuditEvent_DataTest.csv:10]: Sure following that ..

File: ASIM/dev/ASimTester.csv (hunk 1 above)
Comment (rsingh134) [reply ASIM/dev/ASimTester.csv:1]: changes done.

File: Parsers/ASimAuthentication/Parsers/ASimAuthentication.yaml (hunk 8 above)
Comment (rsingh134) [reply Parsers/ASimAuthentication/Parsers/ASimAuthentication.yaml:48]: Change done.

**Pull request closed without merging**